xarray/core/concat.py

```diff
@@ -568,38 +569,40 @@ def get_indexes(name):
                     yield PandasIndex(data, dim, coord_dtype=var.dtype)
 
     # create concatenation index, needed for later reindexing
-    concat_index = list(range(sum(concat_dim_lengths)))
+    concat_index = np.arange(sum(concat_dim_lengths))
+    concat_index_size = concat_index.size
+    variable_index_mask = np.ones(concat_index_size, dtype=bool)
 
     # stack up each variable and/or index to fill-out the dataset (in order)
     # n.b. this loop preserves variable order, needed for groupby.
+    ndatasets = len(datasets)
     for name in vars_order:
         if name in concat_over and name not in result_indexes:
             variables = []
-            variable_index = []
+            mask = variable_index_mask.copy()
```

Comment:
Is it that much  faster to copy than just reassign every time?

Reply:
One of the faster at least, but now that you asked I found `.fill`:
```python
%timeit np.ones(500, dtype=bool)
3.46 µs ± 17.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

a = np.ones(500, dtype=bool)
%timeit a.copy()
429 ns ± 12.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit a[:] = True
450 ns ± 1.56 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit a.fill(False)  # Winner! :)
192 ns ± 2.59 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```